File: Experiment Processing/recommender/distance_metrics/correlation_distance.py
import math
import numpy as np

from recommender.distance_metrics.abstract_distance_metric import AbstractDistanceMetric


class CorrelationDistance(AbstractDistanceMetric):
    """
    Correlation distance metric. See https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.correlation.html
    for more information.
    """

    def get_name(self):
        return "Correlation Distance"

    def calculate_mean(self, track_features):

        mean = 0

        for attribute in self.attribute_keys:
            mean += track_features[attribute]

        return mean / len(self.attribute_keys)

    def calculate_distance(self, track_a, track_b):
        distance_u = 0
        distance_v = 0
        distance_u_v = 0

        track_a_features = track_a.get_metadata()
        track_b_features = track_b.get_metadata()

        mean_a = self.calculate_mean(track_a_features)
        mean_b = self.calculate_mean(track_b_features)

        for attribute in self.attribute_keys:
            attribute_a = track_a_features[attribute]
            attribute_b = track_b_features[attribute]

            distance_u += (attribute_a - mean_a) ** 2
            distance_v += (attribute_b - mean_b) ** 2
            distance_u_v += (attribute_a - mean_a) * (attribute_b - mean_b)

        try:
            distance_u /= len(self.attribute_keys)
            distance_v /= len(self.attribute_keys)
            distance_u_v /= len(self.attribute_keys)
            return 1 - (distance_u_v / math.sqrt(distance_u * distance_v))
        except ZeroDivisionError:
            return np.NaN

File: Web Survey/resources/js/recommender/distance_metrics/CorrelationDistance.js
import AbstractDistanceMetric from "./AbstractDistanceMetric";

export default class CorrelationDistance extends AbstractDistanceMetric {

    static getMetricName() {
        return "Correlation Similarity"
    }

    calculateMean(track, attributes) {
        const self = this;

        let mean = 0;
        this.attributeKeys.forEach(attribute => {
            if(attribute === "loudness") {
                mean += self.fixLoudness(attributes[track][attribute]);
            } else {
                mean += attributes[track][attribute];
            }
        });
        mean /= this.attributeKeys.length;

        return mean;
    }

    calculateDistance(track, otherTrack, attributes) {
        const self = this;

        let distanceU = 0;
        let distanceV = 0;
        let distanceUV = 0;

        let meanTrack = this.calculateMean(track, attributes);
        let meanOther = this.calculateMean(otherTrack, attributes);

        this.attributeKeys.forEach(attribute => {

            let attributeOwn;
            let attributeOther;

            if(attribute === "loudness") {
                attributeOwn = self.fixLoudness(attributes[track][attribute]);
                attributeOther = self.fixLoudness(attributes[otherTrack][attribute])
            } else {
                attributeOwn = attributes[track][attribute];
                attributeOther = attributes[otherTrack][attribute];
            }

            distanceU += (attributeOwn - meanTrack) ** 2;
            distanceV += (attributeOther - meanOther) ** 2;
            distanceUV += (attributeOwn - meanTrack) * (attributeOther - meanOther);
        });
        let distance;

        try {
            distanceUV /= this.attributeKeys.length;
            distanceU /= this.attributeKeys.length;
            distanceV /= this.attributeKeys.length;
            distance = 1 - (distanceUV / Math.sqrt(distanceU * distanceV));
        } catch(err){
            console.log("Zero division");
            return NaN;
        }
        return 1 - distance;
    }

}